SANS TOML V2: Polarization Options Added to User Files

cailafinn opened the pull request:
### Description of work

#### Summary of work
Allows specification of the polarisation options used in the experiment to be provided to the reduction via the user file (a file written in SANS' custom TOML format, usually one per experiment). 

Fixes #38524
 
#### Further detail of work
- Created `StatePolarization`. An JSON-serialisable object that holds the polarisation information to be passed around as part of SANSState to algorithms and interfaces.  
- Refactored the TOML parser into a base and sub-classes to support...
- Implemented SANS TOML V2 parser and schema. Upgrading to V2 means that files with the new information won't be supported by previous versions of Mantid. This is intentional. All V1 files continue to be backwards compatible, all V2 files will warn that the version is unsupported, forcing a version of Mantid to be used that can support the new polarisation options. 
- Refactored some tests to remove duplication caused by the additional parser. 

- Note: If required, this could be 2 PRs. The first implementing the State object and the second implementing the parsing. 

### To test:
1. Download this new user file, modified from the `loqdemo` in the training course data (you'll need that too): [MaskFile_polarization.toml.zip](https://github.com/user-attachments/files/18893469/MaskFile_polarization.toml.zip)
2. Open a nightly and open Interfaces -> SANS -> ISIS SANS
3. Attempt to select the downloaded file as the user file. It should throw. 
4. Open the workbench build of this PR. 
5. Repeat, it should load successfully.
6. Load the batch from the loqdemo
7. Process All
8. Find the output workspaces on the main mantid window and open the workspace history
9. Find the SANSState object
10. Copy it to clipboard, and paste it into a *nice* editor (I like VS code for this)
11. It'll paste in one giant line, reformat the file (install a JSON extension if necessary) 
12. Check that the new polarisation information from the downloaded file is present in the JSON. 


---

### Reviewer


Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

cailafinn committed:
Create StatePolarization Class

RE #38524

cailafinn committed:
Add StateComponent class

RE #38524

cailafinn committed:
Add filter state

RE #38524

cailafinn committed:
Add StateField class

RE #38524

cailafinn committed:
Implement state validation

RE #38524

cailafinn committed:
Add polarization state tests

RE #38524

cailafinn committed:
Implement main polarization parsing

RE #38524

cailafinn committed:
Implement flipper parsing

RE #38524

cailafinn committed:
Parse polarizer and analyzer

RE #38524

cailafinn committed:
Implement field parsing

Covers both the electric and magnetic field inputs.

RE #38524

cailafinn committed:
Add new state to the AllStates Object

RE #38524

cailafinn committed:
Validate the parsed state

RE #38524

cailafinn committed:
Make all parsers produce a polarization object

The old txt format won't be able to edit the values like the toml
parser, but it will return a blank state so that the behaviour of the
two parsers is identical.

RE #38524

cailafinn committed:
Create TOML V2

Given that the new things we're adding to the user file aren't
compatible with older versions of mantid, we're bumping up to TOML V2 so
that we can keep being backwards compatible. This is mainly to stop new
user files containing this information from being loaded by older
versions, making it clear that you need to upgrade to use these new
features and toml files.

RE #38524

cailafinn committed:
Add TOML V2 information to docs

RE #38524

cailafinn committed:
Add a release note

RE #38524

cailafinn commented on scripts/SANS/sans/user_file/txt_parsers/ParsedDictConverter.py:
```diff
@@ -1063,6 +1064,10 @@ def get_state_wavelength_and_pixel_adjustment(self):  # -> StateWavelengthAndPix
         _set_wavelength_limits(state, self._input_dict)
         return state
 
+    def get_state_polarization(self) -> StatePolarization:
```
The `ParsedDictConverter` is a `.txt` parser for the legacy style of user files. Adding this means that it matches the Interface defined in `IStateParser` and the above code using the parser can guarentee a valid sans state with all states created. We don't expect any txt user files to contain polarisation information as that's a brand new set of parameters.

cailafinn commented on scripts/SANS/sans/user_file/toml_parsers/toml_v2_schema.py:
```diff
@@ -0,0 +1,57 @@
+            "electric_field": field_keys,
+        }
+
+        return dict(TomlSchemaV1Validator.reference_schema(), **{"polarization": polarization_keys})
```
I think the order is slightly mixed up. What's the "above" here?

MialLewis replied:
This one: https://github.com/mantidproject/mantid/pull/38940#discussion_r1992118537

cailafinn committed:
Improve dict naming

RE #38524

cailafinn committed:
Stop messing around with __class__

RE #38524

cailafinn committed:
Use match case

RE #38524

cailafinn commented on scripts/SANS/sans/user_file/toml_parsers/toml_base_parser.py:
```diff
@@ -0,0 +1,27 @@
+        self._validator = schema_validator
+        self._validator.validate()
+
+        self._implementation = None
```
I'll find out if it's even necessary. It's taken from the old v1 implementation so it's possible it just got lost in the weeds of the refactor.

cailafinn committed:
Remove now unneeded typing types

RE #38524

cailafinn committed:
Validate substates using vars()

RE #38524

cailafinn commented on scripts/SANS/sans/state/StateObjects/StatePolarization.py:
```diff
@@ -0,0 +1,146 @@
+                    },
+                )
+                is_invalid.update(entry)
+        elif not is_pure_none_or_not_none(direction):
```
Sample directions can be undefined (IDF), defined as log, or as 3 directions. This allows the IDF option to work.

cailafinn committed:
Simplify sample direction validation

RE #38524

cailafinn committed:
Avoid setting implementation twice

RE #38524

cailafinn committed:
Remove TODO

Created an issue instead here: https://github.com/mantidproject/mantid/issues/39103

RE #38524

cailafinn commented on scripts/SANS/sans/test_helper/toml_parser_test_helpers.py:
```diff
@@ -0,0 +1,43 @@
+
+
+def get_mock_data_info():
+    # TODO I really really dislike having to do this in a test, but
```
- https://github.com/mantidproject/mantid/issues/39103

MialLewis reviewed (approved):
Thanks for the changes.

Code looks good and functions as expected.
